### ISSUE TITLE
Is it possible to have screenshot in report?

### ISSUE THREAD

**Issue by dmbfck** [renatoathaydes/spock-reports#117]
My tests extend GebReportingSpec and I'd like to know if there is a way to have the screenshots that GebReportingSpec creates in the spock report?

**Comment by renatoathaydes** [comment]
Can you send me links to the relevant documents explaining how the Geb reporting thing works?

We could add an extension to the html reports to automatically do that if Geb is used... maybe..

**Comment by dmbfck** [comment]
Some documentation on Geb's screenshot (Report) -> http://www.gebish.org/manual/current/#reporting

Not sure how helpful this is but here is a link to GebReportingSpec that I extend for all my tests -> https://github.com/geb/geb/blob/e62c2d52ceafd9bd9601c15c14e3578d93bd7b70/module/geb-spock/src/main/groovy/geb/spock/GebReportingSpec.groovy

Also, if this is something you would be willing to see a PR for, I can try to work on it.  Though, it might take me a little while since this would be the first time doing anything with Spock extensions other than using them.

Thanks!

**Comment by renatoathaydes** [comment]
I think this will be tricky... let me have a look first and if I need help I will let you know.

Thanks.

**Comment by tilmanginzel** [comment]
@renatoathaydes Regarding the geb screenshot integration: More than a year ago, I implemented a Geb screenshot integration into spock-reports using a custom GebReportingListener and a custom spock-reports template. We use that code successfully in a couple of projects. I'm currently in the process of moving that code to a small dedicated (open source) library so it can be more easily shared. 

Now, I would have no problem if we would incorporate that code directly into your library. It would not add too much bloat. If you want to talk a bit about it, you can reach me via github@tilmanginzel.de (or we can just continue in this issue)

**Comment by renatoathaydes** [comment]
Most users of Spock aren't using Geb, I assume, so if this feature is added, it needs to be by a separate module... I am thinking of a service which, if found in the classpath because you added an explicit dependency on it, will attach a "extra" feature into the reports, in this case a screen shot, but that could be whatever...

@tilmanginzel if you could give a hint on how you got that off the ground, at least, it shouldn't be too hard to implement I guess...

**Comment by rdmueller** [comment]
@tilmanginzel sounds interesting - I did the same some time ago. My results can be found here (sorry, mostly german descriptions): https://github.com/rdmueller/etka15

**Comment by tilmanginzel** [comment]
@rdmueller Ah yes, you will see that I shamelessly adopted some of your implementation details. ;)

@renatoathaydes Sure thing. I just made the repository public, you can find it here: [geb-spock-reports](https://github.com/tilmanginzel/geb-spock-reports).

These are the important bits:
1. The [GebReportingListener.groovy](https://github.com/tilmanginzel/geb-spock-reports/blob/master/src/main/groovy/com/aoe/gebspockreport/GebReportingListener.groovy) is called during test execution each time a [Geb report](http://gebish.org/manual/current/#reporting) is called.
2. It writes a `geb-report-info.json` file with meta data about each feature and created screenshots + DOM snapshots. The created file looks currently like this: [geb-report-info.json](https://gist.github.com/tilmanginzel/ebcba1226c3add199c7304e28465e038)
3. I created a custom template based on your HtmlReport: [spec-template](https://github.com/tilmanginzel/geb-spock-reports/blob/master/src/main/resources/templates/spec-template.html). The only addition, besides the complete re-creation of the template, is that it checks if a `geb-report-info.json` is present and it adds links to the screenshots and DOM snapshots at the respective feature.

So tl;dr: one listener and a slightly modified report template.

---

Now, you can think about it if you would like to add this feature directly into your library (or an extra module). If not, I will continue to work on [geb-spock-reports](https://github.com/tilmanginzel/geb-spock-reports) in the coming days and will publish it as soon as I think it is ready.

**Comment by renatoathaydes** [comment]
@tilmanginzel that looks pretty cool! Good work.

I don't want to add Geb support directly into spock-reports, it just needs to be a separate module... but looks like your project already would achieve the same thing...

A small benefit of having a spock-reports "extension" to do this is that it could work for both html reports and all template reports (I would just expose an API to get extra data such as screenshots from templates).

I am currently working on other projects and other features for spock-reports (currently, config improvements and support for displaying which block failed when a spec fails), so let me know when your project is ready and we'll see, maybe having that will be enough for most people.

**Comment by tilmanginzel** [comment]
Alright, fair enough! Then I will continue to work on it and let you guys know when the first version is published.

I will also try to make the template customizable, at least with css files. My plan was also to create a small utility class with access to the geb information, so people can still create custom templates more easily. I'll keep you posted. :+1:

**Comment by rdmueller** [comment]
@tilmanginzel "shamelessly adopted"? I only see that your solution is more elegant - I like it!

**Comment by tilmanginzel** [comment]
Thanks! Just for your information, I have moved the repository [AOEpeople/geb-spock-reports](https://github.com/AOEpeople/geb-spock-reports). I'm planning to publish the first version on bintray in the next week.

**Comment by tilmanginzel** [comment]
I finally found the time to learn about publishing on Bintray/jcenter! :) I'm sorry for the delay, but now it's all set up. @dmbfck Feel free to test if the library works as expected, check the [readme](https://github.com/AOEpeople/geb-spock-reports) for usage instructions.

**Comment by renatoathaydes** [comment]
@tilmanginzel Cool, I will link to this in the spock-reports README and news.

**Comment by tilmanginzel** [comment]
Thank you! :)
